ssl/quic/quic_port.c

```diff
@@ -533,15 +583,29 @@ void ossl_quic_port_drop_incoming(QUIC_PORT *port)
 {
     QUIC_CHANNEL *ch;
     SSL *tls;
+    SSL *user_ssl;
 
     for (;;) {
         ch = ossl_quic_port_pop_incoming(port);
         if (ch == NULL)
             break;
 
         tls = ossl_quic_channel_get0_tls(ch);
-        ossl_quic_channel_free(ch);
-        SSL_free(tls);
+        /*
+         * The user ssl may or may not have been created via the
+         * get_conn_user_ssl callback in the QUIC stack.  The
+         * differentiation being if the user_ssl pointer and tls pointer
+         * are different.  If they are, then the user_ssl needs freeing here
+         * which sends us through ossl_quic_free, which then drops the actual
+         * ch->tls ref and frees the channel
+         */
+        user_ssl = SSL_CONNECTION_GET_USER_SSL(SSL_CONNECTION_FROM_SSL(tls));
+        if (user_ssl == tls) {
```

Comment:
Why would this ever happen?

Reply:
see ossl_ssl_connection_new_int()

If ossl_ssl_connection_new_int is called with user_ssl == NULL, then s->ssl = ssl, and the result of the operation is assigned to ch->tls in port_make_channel.  As such, while it may never happen, its possible that user_ssl == ch->tls and if it does we would call SSL_free on it twice, which would unbalance the refcount

Reply:
> If ossl_ssl_connection_new_int is called with user_ssl == NULL, then s->ssl = ssl

But this is wrong for QUIC!! It never makes sense to do this for the "inner" SSL (TLS) object created as part of QUIC. It only ever makes sense if the application is creating an actual TLS SSL object. If QUIC does this then it will be creating an inner SSL with the user_ssl set to itself which means that the inner SSL will be exposed in TLS callbacks which must never happen.

Looking at the code we seem to do this in `ossl_quic_new_from_listener`. That needs fixing.

Reply:
ok, but thats independent of this PR, and should have a new issue to fix it. We don't need to block this change on such an independent bug.  The check for user_ssl == tls isn't harming anything, and isn't causing the issue you're raising here.

I've opened https://github.com/openssl/project/issues/989 to track the issue